File: portage-ucp-webmcp/spec/support/node_browser.rb
require "json"
require "open3"
require "rack/mock"
require "uri"

# Ruby half of spec/support/node_browser.js: evaluates page scripts in a node
# process standing in for a browser tab, and answers that tab's fetches from
# `rack_app` — the same Rack app a merchant mounts. Specs that need it are
# tagged `:node` and skipped when node isn't installed.
class NodeBrowser
  HARNESS = File.expand_path("node_browser.js", __dir__)
  TIMEOUT = 10

  def self.available?
    system("node", "--version", out: File::NULL, err: File::NULL)
  rescue SystemCallError
    false
  end

  attr_reader :requests

  def initialize(rack_app: nil, origin: "https://shop.example", origin_header: origin)
    @rack_app = rack_app
    @origin = origin
    @origin_header = origin_header
    @requests = []
    @next_id = 0
    @stdin, @stdout, @stderr, @wait = Open3.popen3({ "PAGE_ORIGIN" => origin }, "node", HARNESS)
  end

  # Evaluates `source` in the tab's global scope, awaiting it if it is a
  # promise, and returns what it resolves to.
  def evaluate(source)
    id = (@next_id += 1)
    write(type: "eval", id: id, source: source)
    loop do
      message = read
      case message["type"]
      when "fetch" then answer(message)
      when "result" then return message["value"] if message["id"] == id
      when "error" then raise "page script failed: #{message['message']}" if message["id"] == id
      end
    end
  end

  def close
    @stdin.close unless @stdin.closed?
    @wait.value
  end

  private

  def answer(message)
    @requests << message
    status, body = @rack_app ? rack_response(message) : [503, '{"error":"no rack app"}']
    write(type: "fetch_response", id: message["id"], status: status, body: body)
  end

  def rack_response(message)
    headers = message["headers"].to_h { |key, value| [rack_header(key), value] }
    env = Rack::MockRequest.env_for(URI.join("#{@origin}/", message["url"]).to_s,
                                    method: message["method"], input: message["body"].to_s, **headers)
    env["HTTP_ORIGIN"] = @origin_header if @origin_header
    status, _headers, body = @rack_app.call(env)
    [status, body.enum_for(:each).to_a.join]
  end

  def rack_header(name)
    return "CONTENT_TYPE" if name.casecmp?("content-type")

    "HTTP_#{name.upcase.tr('-', '_')}"
  end

  def write(message)
    @stdin.puts(JSON.generate(message))
    @stdin.flush
  end

  def read
    raise "node browser timed out: #{drain_stderr}" unless @stdout.wait_readable(TIMEOUT)

    line = @stdout.gets
    raise "node browser exited: #{drain_stderr}" unless line

    JSON.parse(line)
  end

  def drain_stderr
    @stderr.wait_readable(0) ? @stderr.read_nonblock(65_536, exception: false).to_s : ""
  end
end

File: portage-ucp-webmcp/spec/support/node_browser.js
// Stand-in for a browser tab, driven by spec/support/node_browser.rb over
// newline-delimited JSON on stdin/stdout. It gives the page scripts exactly
// what they touch in a browser — window/document/navigator/location and
// fetch — and nothing else, so registrar.js, consumer.js and polyfill.js run
// unmodified. `fetch` is relayed back to Ruby, which answers it from a Rack
// app, so a tool call really does travel page -> endpoint -> Mcp::Server.
"use strict";

const readline = require("readline");

globalThis.window = globalThis;
globalThis.document = {};
globalThis.location = { origin: process.env.PAGE_ORIGIN || "https://shop.example" };
if (typeof globalThis.navigator === "undefined") globalThis.navigator = {};

const send = (message) => process.stdout.write(JSON.stringify(message) + "\n");
const pending = new Map();
let nextFetch = 0;

globalThis.fetch = (url, init = {}) => {
  const id = ++nextFetch;
  send({
    type: "fetch", id, url: String(url), method: init.method || "GET",
    headers: init.headers || {}, credentials: init.credentials || null, body: init.body || null
  });
  return new Promise((resolve) => pending.set(id, resolve)).then((response) => ({
    status: response.status,
    ok: response.status >= 200 && response.status < 300,
    json: () => Promise.resolve(JSON.parse(response.body)),
    text: () => Promise.resolve(response.body)
  }));
};

readline.createInterface({ input: process.stdin }).on("line", (line) => {
  const message = JSON.parse(line);
  if (message.type === "fetch_response") {
    pending.get(message.id)(message);
    pending.delete(message.id);
    return;
  }
  if (message.type === "eval") {
    Promise.resolve()
      .then(() => (0, eval)(message.source))
      .then(
        (value) => send({ type: "result", id: message.id, value: value === undefined ? null : value }),
        (error) => send({ type: "error", id: message.id, message: String((error && error.stack) || error) })
      );
  }
});
